### PR TITLE
ci: Apply security best practices based on StepSecurity recommendations

### PR DESCRIPTION
**Pull request by tippmar-nr**
## TL;DR
* Applied StepSecurity best practices to our workflows:
    * Converted version references to commit SHA references for all actions referenced in our workflows
    * Added the `harden-runner` action to all Ubuntu-based workflow jobs, running in "audit" mode where recommendations will be made for future configuration changes for the `harden-runner` action. A follow-up task will need to be done in a week or two, to apply recommendations gathering during the audit phase.
    * ~Added a `dependency-review` workflow that will scan for known vulnerabilities in NuGet packages.~ 
    * Added a `dependabot.yml` file to configure Dependabot to scan for updated actions ~as well as NuGet packages~. Note that update scanning is presently disabled for our repo. We would need to turn it on before this file has any effect.
    * Added `permissions:` tag to most workflow jobs to limit the permissions of the Github token to the least level required.
* Note that the original PR also included a workflow for running CodeQL static analysis, but I removed it, as it wouldn't run against our C# or C++ code without a lot more configuration than I was interested in messing with.

The original PR text from StepSecurity follows:

## Summary

This pull request is created by [Secure Repo](https://app.stepsecurity.io/securerepo) at the request of @tippmar-nr. Please merge the Pull Request to incorporate the requested changes. Please tag @tippmar-nr on your message if you have any questions related to the PR. You can also engage with the [StepSecurity](https://github.com/step-security) team by tagging @step-security-bot.

### Least Privileged GitHub Actions Token Permissions

The GITHUB_TOKEN is an automatically generated secret to make authenticated calls to the GitHub API. GitHub recommends setting minimum token permissions for the GITHUB_TOKEN.

- [GitHub Security Guide](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#using-the-github_token-in-a-workflow)
- [The Open Source Security Foundation (OpenSSF) Security Guide](https://github.com/ossf/scorecard/blob/main/docs/checks.md#token-permissions)
### Pinned Dependencies

GitHub Action tags and Docker tags are mutatble. This poses a security risk. GitHub's Security Hardening guide recommends pinning actions to full length commit.

- [GitHub Security Guide](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions)
- [The Open Source Security Foundation (OpenSSF) Security Guide](https://github.com/ossf/scorecard/blob/main/docs/checks.md#pinned-dependencies)
### Harden Runner

[Harden-Runner](https://github.com/step-security/harden-runner) is an open-source security agent for the GitHub-hosted runner to prevent software supply chain attacks. It prevents exfiltration of credentials, detects tampering of source code during build, and enables running jobs without `sudo` access.

<details>
<summary>Harden runner usage</summary>

You can find link to view insights and policy recommendation in the build log

Please refer to [documentation](https://docs.stepsecurity.io/harden-runner/getting-started#why-monitor-cicd-runtime) to find more details.
</details>

### Keeping your actions up to date with Dependabot

With Dependabot version updates, when Dependabot identifies an outdated dependency, it raises a pull request to update the manifest to the latest version of the dependency. This is recommended by GitHub as well as The Open Source Security Foundation (OpenSSF).

- [GitHub Security Guide](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/keeping-your-actions-up-to-date-with-dependabot)
- [The Open Source Security Foundation (OpenSSF) Security Guide](https://github.com/ossf/scorecard/blob/main/docs/checks.md#dependency-update-tool)
### Detect Vulnerabilities with SAST Workflow 

Static Code Analysis (also known as Source Code Analysis) is usually performed as part of a Code Review (also known as clear-box testing) and is carried out at the Implementation phase of a Security Development Lifecycle (SDL). Static Code Analysis commonly refers to the running of Static Code Analysis tools that attempt to highlight possible vulnerabilities within ‘static’ (non-running) source code by using techniques such as Taint Analysis and Data Flow Analysis.

- [The Open Source Security Foundation (OpenSSF) Security Guide](https://github.com/ossf/scorecard/blob/main/docs/checks.md#sast)
- [OWASP Static Code Analysis](https://owasp.org/www-community/controls/Static_Code_Analysis)
- [Github Guide For Code Scanning](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/setting-up-code-scanning-for-a-repository)
### Add Dependency Review Workflow

The Dependency Review Workflow enforces dependency reviews on your pull requests. The action scans for vulnerable versions of dependencies introduced by package version changes in pull requests, and warns you about the associated security vulnerabilities. This gives you better visibility of what's changing in a pull request, and helps prevent vulnerabilities being added to your repository.

- [Github Guide about Dependency Review](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/about-dependency-review)
- [Github Guide for Configuring Dependency Review Action](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/configuring-dependency-review#using-inline-configuration-to-set-up-the-dependency-review-action)

## Feedback
For bug reports, feature requests, and general feedback; please create an issue in [step-security/secure-repo](https://github.com/step-security/secure-repo). To create such PRs, please visit https://app.stepsecurity.io/securerepo.

Signed-off-by: StepSecurity Bot <bot@stepsecurity.io>

### CODE REVIEW

File: .github/workflows/dependency-review.yml
Comment (chynesNR) [inline .github/workflows/dependency-review.yml:1]: Won't this cause a problem with our integration tests?

File: .github/dependabot.yml
Comment (nrcventura) [inline .github/dependabot.yml:1]: I think that we should consider pulling this dependabot configuration out of this PR or reducing it significantly. I think that we should be more selective about what we enable dependabot for. 
For example, I think that dependabot can be helpful for updating github action dependencies, some docker containers, and some nuget dependencies. However, the current version of this file will trigger notifications about newer package versions for things that we do not want to update. For example, it will prompt us to update the dependencies of our instrumentation projects which are purposely set to specific older versions. A similar thing will happen to our test app projects where specific versions were chosen to not be the latest version.

File: .github/workflows/dependency-review.yml
Comment (tippmar-nr) [reply .github/workflows/dependency-review.yml:1]: Actually, it turns out that this workflow is for NPM type packages - running it on our repo produced no warnings or errors of any kind. So I removed it.

File: .github/workflows/assignproj.yml [hunk 1]
```diff
@@ -7,14 +7,22 @@ on:
...
+
 jobs:
   assign_one_project:
     runs-on: ubuntu-latest
     name: Assign to One Project
     steps:
+    - name: Harden Runner
+      uses: step-security/harden-runner@03bee3930647ebbf994244c21ddbc0d4933aab4f # v2.3.0
+      with:
+        egress-policy: audit # TODO: change to 'egress-policy: block' after couple of runs
```
Comment (chynesNR) [inline .github/workflows/assignproj.yml:21]: Could you create a ticket to address the TODOs?

**Review by chynesNR — APPROVED** [review]

File: .github/workflows/assignproj.yml (hunk 1 above)
Comment (tippmar-nr) [reply .github/workflows/assignproj.yml:21]: Done. #1515

Comment (chynesNR) [reply .github/workflows/assignproj.yml:21]: Thanks!

**Review by nrcventura — APPROVED** [review]
